File: backend/app/core/phone.py
"""หมายเลขโทรศัพท์ไทย — ทำให้อยู่รูปแบบเดียวกันก่อนเก็บและก่อนค้นหา

ทำไมต้อง normalize ก่อนเก็บ:
M1 ให้เข้าสู่ระบบด้วยเบอร์โทรได้ ถ้าผู้ใช้สมัครด้วย "081-234-5678"
แล้วพิมพ์ "0812345678" ตอนเข้าสู่ระบบ ระบบต้องรู้ว่าเป็นคนเดียวกัน
จึงเก็บเป็นตัวเลขล้วนเสมอ แล้วค่อยใส่ขีดตอนแสดงผล

ตรรกะเดียวกับ frontend/src/lib/phone.ts — แก้ไฟล์ใดต้องแก้อีกไฟล์ด้วย
"""

MOBILE_LENGTH = 10  # 08x-xxx-xxxx
LANDLINE_LENGTH = 9  # 076-xxx-xxx


def normalize(value: str) -> str:
    """เหลือเฉพาะตัวเลข และแปลงรูปแบบสากล (+66 81 234 5678) ให้เป็น 0812345678

    รับรูปแบบสากลด้วยเพราะผู้ประกอบการที่เคยลงทะเบียนกับแพลตฟอร์มจองที่พัก
    มักคุ้นกับการเขียนเบอร์แบบ +66
    """
    digits = "".join(ch for ch in value if ch.isdigit())
    if digits.startswith("66") and len(digits) in (LANDLINE_LENGTH + 1, MOBILE_LENGTH + 1):
        digits = "0" + digits[2:]
    return digits


def is_valid(value: str) -> bool:
    """เบอร์มือถือ 10 หลัก หรือเบอร์บ้าน 9 หลัก และต้องขึ้นต้นด้วย 0

    ไม่ตรวจละเอียดกว่านี้ (เช่น บังคับขึ้นต้น 06/08/09) เพราะช่วงหมายเลข
    ที่ กสทช. จัดสรรเปลี่ยนได้ การบล็อกเกินจำเป็นจะกันผู้ใช้จริงออกไป
    """
    digits = normalize(value)
    return len(digits) in (LANDLINE_LENGTH, MOBILE_LENGTH) and digits.startswith("0")


def format_display(value: str) -> str:
    """ใส่ขีดให้อ่านง่าย -> 081-234-5678 หรือ 076-123-456"""
    d = normalize(value)
    if len(d) not in (LANDLINE_LENGTH, MOBILE_LENGTH):
        return value
    return f"{d[:3]}-{d[3:6]}-{d[6:]}"

File: frontend/src/lib/phone.ts
/**
 * หมายเลขโทรศัพท์ไทย — ฝั่งหน้าเว็บ
 *
 * ตรรกะเดียวกับ backend/app/core/phone.py
 * ทำซ้ำฝั่งนี้เพื่อให้ผู้ใช้รู้ผลทันทีที่พิมพ์ครบ ไม่ต้องรอส่งฟอร์ม
 * แต่ backend ยังต้องตรวจซ้ำเสมอ เพราะการตรวจฝั่งหน้าเว็บถูกข้ามได้
 *
 * ถ้าแก้สูตรที่ไฟล์ใดไฟล์หนึ่ง ต้องแก้อีกไฟล์ด้วย
 */

/** ความยาวสูงสุดที่ยอมรับ — ใช้ตัดค่าที่ช่องกรอกด้วย ไม่งั้นผู้ใช้พิมพ์เกิน
    แล้วเห็นบนจอ 10 หลักถูกต้อง แต่ค่าในฟอร์มมี 11 หลักและถูกตีว่าผิด */
export const PHONE_MAX_DIGITS = 10;

const MOBILE_LENGTH = PHONE_MAX_DIGITS; // 08x-xxx-xxxx
const LANDLINE_LENGTH = 9; // 076-xxx-xxx

/** เหลือเฉพาะตัวเลข และแปลงรูปแบบสากล (+66 81 234 5678) ให้เป็น 0812345678 */
export function normalizePhone(value: string): string {
  const digits = value.replace(/\D/g, "");
  if (digits.startsWith("66") && (digits.length === MOBILE_LENGTH + 1 || digits.length === LANDLINE_LENGTH + 1)) {
    return `0${digits.slice(2)}`;
  }
  return digits;
}

/** เบอร์มือถือ 10 หลัก หรือเบอร์บ้าน 9 หลัก และต้องขึ้นต้นด้วย 0 */
export function isValidPhone(value: string): boolean {
  const d = normalizePhone(value);
  return (d.length === MOBILE_LENGTH || d.length === LANDLINE_LENGTH) && d.startsWith("0");
}

/** จัดรูปแบบระหว่างพิมพ์ -> 081-234-5678 */
export function formatPhone(value: string): string {
  const d = normalizePhone(value).slice(0, MOBILE_LENGTH);
  return [d.slice(0, 3), d.slice(3, 6), d.slice(6)].filter(Boolean).join("-");
}
